### PR TITLE
Batch Job name

### LINKED ISSUE #15: unexpected keyword
We have 0.5.3 and on the anaconda.org channel, that is the latest one. 

 
```
@tranquilize(method='GET')

def submit():

    kinit_username(user, pw)

    file_on_hdfs = upload(LOCAL_FILE, namenode_url=namenode_url)

    livy_api = LivyAPI(server_url=livy_url)

    batch = livy_api.submit(

                            file=file_on_hdfs, name='livy test')

    return json.dumps(repr(batch))
```
 

1. When executing submit(), this ran and placed the local_file in `/user/<user ID>/livy-submit-files/<random folder>/<codefile>.py`. This works and I can see the file in the cluster.

2. When executing submit(), it threw the following error

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-7e910993d2f2> in <module>()
      1 # For debugging, we'll want to run this submit function to make sure our code works
----> 2 submit()
 
<ipython-input-4-4f5d153e388e> in submit()
      5     livy_api = LivyAPI(server_url=livy_url)
      6     batch = livy_api.submit( 
----> 7                             file=file_on_hdfs, name='livy test')
      8     return json.dumps(repr(batch))
 
~/anaconda/envs/anaconda50_hadoop/lib/python3.6/site-packages/livy_submit/livy_api.py in submit(self, name, file, driverMemory, driverCores, executorMemory, executorCores, numExecutors, archives, queue, conf, args, pyFiles)
    197         # Submit the data dict to the Livy Batches API to create a batch job
    198         response = self._request("post", self._base_url, data=data)
--> 199         return Batch(**response)
    200 
    201     def log(
 
TypeError: __init__() got an unexpected keyword argument 'name'
 ```

However on Livy, I noticed that this actually ran successfully. Note – The session was still getting started when the notebook threw the error.

 

3. I reran the notebook and this time, it failed with the below error and there were no new Livy sessions created.

```
---------------------------------------------------------------------------
HTTPError                                 Traceback (most recent call last)
<ipython-input-12-7e910993d2f2> in <module>()
      1 # For debugging, we'll want to run this submit function to make sure our code works
----> 2 submit()
 
<ipython-input-11-4f5d153e388e> in submit()
      5     livy_api = LivyAPI(server_url=livy_url)
      6     batch = livy_api.submit( 
----> 7                             file=file_on_hdfs, name='livy test')
      8     return json.dumps(repr(batch))
 
~/anaconda/envs/anaconda50_hadoop/lib/python3.6/site-packages/livy_submit/livy_api.py in submit(self, name, file, driverMemory, driverCores, executorMemory, executorCores, numExecutors, archives, queue, conf, args, pyFiles)
    196         #         print(data)
    197         # Submit the data dict to the Livy Batches API to create a batch job
--> 198         response = self._request("post", self._base_url, data=data)
    199         return Batch(**response)
    200 
 
~/anaconda/envs/anaconda50_hadoop/lib/python3.6/site-packages/livy_submit/livy_api.py in _request(self, rest_action, url, data)
    293         )
    294         # Make sure that our request was successful
--> 295         resp.raise_for_status()
    296         return resp.json()
 
~/anaconda/envs/anaconda50_hadoop/lib/python3.6/site-packages/requests/models.py in raise_for_status(self)
    933 
    934         if http_error_msg:
--> 935             raise HTTPError(http_error_msg, response=self)
    936 
    937     def close(self):
 
HTTPError: 400 Client Error: Bad Request for url: http://<url>:8998/batches
```

### PR DESCRIPTION
**Pull request by AlbertDeFusco**
This fixes #15 
* `Batch` now support `name` as an argument in `__init__`
* `submit()` will throw `ValueError` if the name of the submitted job matches one returned by `all_info()`
    * test added

### CODE REVIEW

**Comment by AlbertDeFusco** [comment]
Actually, I need to check if `name` is actually required or can be left off. This fix assumes that you must supply `name` and that it is unique.

**Comment by AlbertDeFusco** [comment]
The job name is now optional in the CLI and submit function

File: livy_submit/livy_api.py [hunk 1]
```diff
@@ -195,6 +197,19 @@ def submit(
                 continue
             data[var] = val
         #         print(data)
+
+        # check for collision with existing batch name
+        if name is not None:
+            _, _, batches = self.all_info()
+            for num,batch in batches.items():
+                if batch.name == name:
```
Comment (ericdill) [inline livy_submit/livy_api.py]: is it possible that `batch.name` here could be `None`?

File: livy_submit/cli.py [hunk 2]
```diff
@@ -403,7 +403,7 @@ def _livy_submit_parser(subparsers):
     ap.add_argument(
         "--name",
         action="store",
-        required=True,
+        required=False,
         help="The name that your Spark job should have on the Yarn RM",
```
Comment (ericdill) [inline livy_submit/cli.py]: What does the Yarn RM say is the "Name" if we don't provide a name? I'd like to document that so we know for future us. Whatever we discover should also be documented in the LivyAPI.submit() function.

File: livy_submit/livy_api.py (hunk 1 above)
Comment (AlbertDeFusco) [reply livy_submit/livy_api.py]: Yes, I'll adjust to `if (batch.name is not None) and (batch.name == name)`

File: livy_submit/cli.py (hunk 2 above)
Comment (AlbertDeFusco) [reply livy_submit/cli.py]: on the resource manager the Name is given as the name of the file and can appear multiple times. From the LivyAPI name will return as JSON `null`, which is parsed to the singleton `None`.

Comment (AlbertDeFusco) [reply livy_submit/cli.py]: added to docstring

**Pull request merged**